instrumentation/github.com/confluentinc/confluent-kafka-go/kafka/splunkkafka/producer.go

```diff
@@ -0,0 +1,148 @@
+}
+
+// NewProducer calls kafka.NewProducer and wraps the resulting Producer with
+// tracing instrumentation.
+func NewProducer(conf *kafka.ConfigMap, opts ...Option) (*Producer, error) {
+	p, err := kafka.NewProducer(conf)
+	if err != nil {
+		return nil, err
+	}
+	return WrapProducer(p, opts...), nil
+}
+
+// WrapProducer wraps a kafka.Producer so that any produced events are traced.
+func WrapProducer(p *kafka.Producer, opts ...Option) *Producer {
+	wrapped := &Producer{
+		Producer: p,
+		cfg:      newConfig(opts...),
+	}
+	wrapped.produceChannel = wrapped.traceProduceChannel(p.ProduceChannel())
+	return wrapped
+}
+
+func (p *Producer) traceProduceChannel(out chan *kafka.Message) chan *kafka.Message {
+	if out == nil {
+		return nil
+	}
+
+	in := make(chan *kafka.Message, cap(out))
+	go func() {
+		for msg := range in {
+			span := p.startSpan(msg)
+			out <- msg
+			span.End()
+		}
+	}()
+
+	return in
+}
+
+func (p *Producer) startSpan(msg *kafka.Message) trace.Span {
+	carrier := NewMessageCarrier(msg)
+	psc := p.cfg.Propagator.Extract(context.Background(), carrier)
+
+	attrs := []attribute.KeyValue{
+		semconv.MessagingSystemKey.String("kafka"),
+		semconv.MessagingDestinationKindTopic,
+		semconv.MessagingDestinationKey.String(*msg.TopicPartition.Topic),
+		semconv.MessagingMessageIDKey.String(strconv.FormatInt(int64(msg.TopicPartition.Offset), 10)), // nolint: gomnd
+		semconv.MessagingKafkaMessageKeyKey.String(string(msg.Key)),
+		semconv.MessagingKafkaPartitionKey.Int64(int64(msg.TopicPartition.Partition)),
+	}
+	opts := []trace.SpanStartOption{
+		trace.WithAttributes(attrs...),
+		trace.WithSpanKind(trace.SpanKindProducer),
+	}
+
+	name := fmt.Sprintf("%s send", *msg.TopicPartition.Topic)
+	ctx, span := p.cfg.Tracer.Start(psc, name, opts...)
+	// Inject the current span into the original message so it can be used to
+	// propagate the span.
+	p.cfg.Propagator.Inject(ctx, carrier)
+	return span
+}
+
+// Close calls the wrapped Producer.Close and closes the producer channel.
+func (p *Producer) Close() {
+	close(p.produceChannel)
+	p.Producer.Close()
+}
+
+// Produce calls the wrapped Producer.Produce and traces the request.
+func (p *Producer) Produce(msg *kafka.Message, deliveryChan chan kafka.Event) error {
+	span := p.startSpan(msg)
+
+	// if the user has selected a delivery channel, we will wrap it and
+	// wait for the delivery event to finish the span
+	if deliveryChan != nil {
+		oldDeliveryChan := deliveryChan
+		deliveryChan = make(chan kafka.Event)
+		go func() {
+			var err error
+			evt := <-deliveryChan
+			if respMsg, ok := evt.(*kafka.Message); ok {
+				// delivery errors are returned via TopicPartition.Error
+				err = respMsg.TopicPartition.Error
+			}
+			if err != nil {
+				span.RecordError(err)
+				span.SetStatus(codes.Error, err.Error())
+			}
+			span.End()
+			oldDeliveryChan <- evt
+		}()
+	}
+
+	err := p.Producer.Produce(msg, deliveryChan)
+	// with no delivery channel, finish immediately
```

Comment:
shouldn't we wrap `Events()` to handle this scenario? I think it is a quite regular use case just to do "fire-and-foget"

See: https://pkg.go.dev/github.com/confluentinc/confluent-kafka-go@v1.7.0/kafka?utm_source=gopls#Producer.Produce

Reply:
You can create an issue out of it if you do not want to spend time on this right now.

Reply:
I'm not sure it is possible or maybe just not worth it. If there is no delivery channel we need to track the state of the enqueued message. Otherwise, we will not be able to ensure the span ended corresponds to the message we just enqueued. Given we do not know the offset of the message when we enqueue the message here, I'm not sure we could ever actually re-associate the span we create here with the response message.

It might be possible because the underlying implementation looks to return the same pointer to the enqueued message, but there is no guarantee this implementation will continue.

I think the best we can do is what is done here - end the span once we have successfully enqueued it.

Feel free to open an issue to track this if you think there is indeed a way to handle this.

Reply:
Looking a the existing solution, I think we suffer from the same issue when the user provides a delivery channel. There is no guarantee they do not provide that channel to multiple calls to `Produce`. Meaning that our assumed guarantee that the response returned on that deliver channel is going to be associated with that specific call is incorrect at times.